### PR TITLE
handle syntax errors more gracefully

### PR DESCRIPTION
**Pull request by aaronjanse**
Fixes https://github.com/nix-community/rnix-parser/issues/33.

I don't know enough about `rowan` to cleanly improve error handling, but I thought I'd at least share the patch I've been using locally while implementing rnix-lsp autocomplete.

Feel free to re-use parts of this patch for a different PR, push directly to this PR, or something else.

### CODE REVIEW

File: src/parser.rs [hunk 1]
```diff
@@ -291,7 +301,15 @@ where
...
             Some(TOKEN_DYNAMIC_START) => self.parse_dynamic(),
             Some(TOKEN_STRING_START) => self.parse_string(),
-            _ => self.expect_ident(),
+            Some(TOKEN_IDENT) => self.expect_ident(),
+            _ => {
+                let start = self.start_error_node();
+                let end = self.finish_error_node();
+                self.errors.push(ParseError::Expected(
+                    TextRange::new(start, end),
+                    "attribute".to_string(),
```
Comment (Ma27) [inline src/parser.rs:310]: I guess we should reflect here that strings etc. are also OK.

**Review by Ma27 — COMMENTED** [review]
LGTM mostly, nice work! :) 

Would you mind resolving the last comment and rebase?

File: src/parser.rs (hunk 1 above)
Comment (aaronjanse) [reply src/parser.rs:310]: What should be the exact message used? Maybe `attribute or string`? Although I guess a string does represent an attribute here? Please correct me if I'm wrong

**Comment by aaronjanse** [comment]
I pushed a somewhat significant change to behavior. The change matters only for attr path now. I also added an "expected semicolon" error node and updated tests

**Comment by oberblastmeister** [comment]
You need to run this with a fuzzer. Adding better error handling can easily screw up the parser, and I got an out of memory error relatively quickly. I would prefer more general error handling (removing `expect_peek_any`). It would be also be good to see the effectiveness of the error handling by hooking it up with `rnix-lsp`. Maybe a mono repo would be good for this?

**Comment by aaronjanse** [comment]
Wow, good catch @oberblastmeister. The fuzzer is amazing, and after seeings its result, I agree that we need a more general  solution here. I'll start working on that. Thank you!!

**Comment by oberblastmeister** [comment]
@aaronjanse I already did a lot of work here https://github.com/nix-community/rnix-parser/pull/35. Also, I really think we should work on hooking up the new version of `rnix` to `rnix-lsp` before we do improved error handling. This way we can try out the error handling in real time.

**Comment by aaronjanse** [comment]
> @aaronjanse I already did a lot of work here https://github.com/nix-community/rnix-parser/pull/35

Oh, sorry about that! I'm still catching up after being away for a while. If you submit a PR for the error handling that was in #35, I'll close #120 in favor of your PR.

> I really think we should work on hooking up the new version of rnix to rnix-lsp before we do improved error handling. This way we can try out the error handling in real time.

As in, you think we should update rnix-lsp to use the latest nix-parser version? I'd be willing to work on doing that

**Comment by oberblastmeister** [comment]
> As in, you think we should update rnix-lsp to use the latest nix-parser version? I'd be willing to work on doing that

Yes! I think that would be great.

**Comment by aaronjanse** [comment]
Sounds good, will do!

**Pull request closed without merging**